File: content/blogs/the-pi-of-archimedes/auxiliary/pi_approximation.py
#!/usr/bin/env python3
"""Approximate π."""
import math

from texttable import Texttable

# Number of sides of the polygon inscribing and circumscribing the circle
NUM_SIDES = [6, 12, 24, 48, 96, 100, 1000, 10000, 100000, 1000000]


def main():
    """Approximate π using an inscribed and circumscribed n-sided polygon."""
    pi_approximations = [["n", "n sin π/n", "n tan π/n"]]
    for n in NUM_SIDES:
        # Calculate the approximation for π using the inscribed circle
        sin_approximation = round(n * math.sin(math.pi / n), 10)

        # Calculate the approximation for π using the circumscribed circle
        tan_approximation = round(n * math.tan(math.pi / n), 10)

        # Append a list containing containing the number of sides (n)
        # and the sine and tan approximations to a list
        pi_approximations.append(
            [
                f"{n}",
                f"{sin_approximation:1.10f}",
                f"{tan_approximation:1.10f}",
            ]
        )

    print_table(pi_approximations)


def print_table(pi_approximations):
    """Print a text table containing the approximations."""
    table = Texttable()

    table.set_cols_align(["r", "r", "r"])
    table.set_header_align(["r", "r", "r"])
    table.set_cols_valign(["b", "b", "b"])
    table.set_cols_dtype(["i", "f", "f"])
    table.set_precision(10)  # Prevent truncation of floats
    table.add_rows(pi_approximations)

    print(table.draw())


if __name__ == "__main__":
    main()

File: content/blogs/the-pi-of-archimedes/auxiliary/pi_approximation.jl
#!/usr/bin/env -S julia -t 4
#
# Author: Nandakumar Chandrasekhar
# Date: 10 July 2024
#
# Approximate π to 10 decimal places by increasing the number
# of sides on an n-sided polygon that is inscribing and circumscribing a circle
using PrettyTables

# Number of sides of the polygon inscribing and circumscribing the circle
const NUM_SIDES = [6, 12, 24, 48, 96, 100, 1000, 10000, 100000, 1000000]

function main()
    # Initialize a Float Matrix with unlimited rows and 3 columns
    pi_approximations = Array{Float64}(undef, 0, 3)

    for n in NUM_SIDES
        # Calculate the approximation for π using the inscribed circle
        sin_approximation = round(n * sin(π / n), sigdigits=11)

        # Calculate the approximation for π using the circumscribed circle
        tan_approximation = round(n * tan(π / n), sigdigits=11)

        # Append a vector containing the number of sides (n)
        # and the sine and tan approximations
        pi_approximations = vcat(
            pi_approximations, reshape([n, sin_approximation, tan_approximation], 1, 3),
        )
    end

    print_table(pi_approximations)
end

function print_table(pi_approximations)
    # Highlight the side column
    hl_side_col = Highlighter(
        f=(data, i, j) -> j == 1,
        crayon=Crayon(foreground=:green),
    )

    # Hightlight the sine approximation column
    hl_sin_col = Highlighter(
        f=(data, i, j) -> j == 2,
        crayon=Crayon(foreground=:cyan),
    )

    # Highlight the tan approximation column
    hl_tan_col = Highlighter(
        f=(data, i, j) -> j == 3,
        crayon=Crayon(foreground=:magenta),
    )

    # Print a table containing the approximations for π
    pretty_table(
        pi_approximations;
        formatters=(ft_printf("%d", 1), ft_printf("%11.10f", 2:3)), # Format specifiers
        header=["n", "n sin π/n", "n tan π/n"],
        highlighters=(hl_side_col, hl_sin_col, hl_tan_col), # Column value color
        header_crayon=Crayon(foreground=:yellow, bold=:true), # Header row color
        crop=:none, # Stop pretty_table from truncating rows
        tf=tf_markdown,
    )
end

main()
